### DISCUSSION TITLE
Analyzing LLVM IR version > 10

### DISCUSSION THREAD

**Discussion by nano-o** [seahorn/seahorn#448] [Q&A]
Hi all, I would like to try seahorn on a Rust project which I can only compile to LLVM IR version 12 or above. I presume this is going to be a problem for seahorn since it uses LLVM 10. Is that so? If it is a problem, can it be solved with a few hacks or would it require more significant engineering effort?

**Comment by caballa** [comment]
Moving to more recent LLVM versions is on our TODO list but unfortunately we haven't had time for that.
The way to go is to port from LLVM 11 first and then to LLVM 12. 

SeaHorn depends on three external packages:
- [llvm-seahorn](https://github.com/seahorn/llvm-seahorn)
- [clam](https://github.com/seahorn/clam)
- [sea-dsa](https://github.com/seahorn/sea-dsa)

Clam and sea-dsa have been ported already to LLVM 12 (see [dev12](https://github.com/seahorn/clam/tree/dev12) and  [dev12](https://github.com/seahorn/sea-dsa/tree/dev12)).  llvm-seahorn has been ported only to LLVM 11 ([dev11](https://github.com/seahorn/llvm-seahorn/tree/dev11)) but llvm-seahorn is optional so you can skip it for now.

To port SeaHorn to llvm 12 shouldn't be hard. There is already a branch [dev10-no-callsite](https://github.com/seahorn/seahorn/tree/dev10-no-callsite) that was created as an intermediate step while porting to LLVM 11. 

I would take the branch `dev10-no-callsite` and port it to LLVM 11. You can also see the commit changes for clam and sea-dsa when they were ported to LLVM 11 and 12 to see which changes were made to guide you here.  I would create a new branch called `dev11` (I can create the branch for you and then you just make a PR) and then from there do the same and port to LLVM 12 creating a new branch `dev12`. The key is to be incremental. If you try to jump too many LLVM versions porting will be hard. 

Let me know if you want to help. I'm happy to guide you through the process. For instance, the first thing to do would be to merge `dev10` into `dev10-no-callsite` but I can do that.

**Reply by nano-o** [reply]
Thanks! It seems that it would take me some time to get familiar with the project before I can do something like that. I will come back to this if lower-hanging fruits don't pan out (e.g. Klee seems to support LLVM up to 13).

**Comment by priyasiddharth** [comment]
Hello,
If you just want to try SEAHORN (BMC engine) on Rust you can take a look at SEAURCHIN. We package Rust `1.46` in the tool so it works with LLVM 10.

https://github.com/priyasiddharth/seaurchin

We have plans to support LLVM 12 and higher in the near future.

**Reply by nano-o** [reply]
Thanks! Unfortunately one of my constraints is to use Rust 1.63 (the project I'm working uses all the latest Rust features...)

**Comment by priyasiddharth** [comment]
Blocked on [449](https://github.com/seahorn/seahorn/issues/449)
